locations/middlewares/anti_bot_detection.py

```diff
@@ -0,0 +1,86 @@
+# Zyte API's ability to rapidly bypass new/revised anti-bot methods.
+class AntiBotMethods(Enum):
+    AZURE_WAF = {"name": "Azure WAF", "zyte_bypassable": True}
+    CLOUDFLARE = {"name": "Cloudflare", "zyte_bypassable": True}
+    DATADOME = {"name": "DataDome", "zyte_bypassable": True}
+    HUMAN = {"name": "HUMAN", "zyte_bypassable": True}
+    IMPERVA = {"name": "Imperva", "zyte_bypassable": True}
+    QRATOR = {"name": "Qrator", "zyte_bypassable": True}
+
+
+class AntiBotDetectionMiddleware:
+    def __init__(self, crawler):
+        self.crawler = crawler
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        return cls(crawler)
+
+    def process_response(self, request: Request, response: Response, spider: Spider):
+        if response.headers.get("cf-mitigated"):
+            self.add_anti_bot_method(AntiBotMethods.CLOUDFLARE, spider)
+
+        if response.headers.get("x-datadome"):
+            self.add_anti_bot_method(AntiBotMethods.DATADOME, spider)
+
+        if server := response.headers.get("server"):
+            server_utf8 = self.decode_http_header_value(server)
+            if server_utf8.upper() == "QRATOR":
+                self.add_anti_bot_method(AntiBotMethods.QRATOR, spider)
+            if response.status == 403 and server_utf8.upper() == "MICROSOFT-AZURE-APPLICATION-GATEWAY/V2":
+                self.add_anti_bot_method(AntiBotMethods.AZURE_WAF, spider)
+
+        if cookies := response.headers.getlist("set-cookie"):
+            for cookie in cookies:
+                cookie_utf8 = self.decode_http_header_value(cookie)
+                if cookie_utf8.startswith("visid_incap_") or cookie_utf8.startswith("incap_ses_"):
+                    self.add_anti_bot_method(AntiBotMethods.IMPERVA, spider)
+
+        if response.status == 403:
+            # We may need to decompress the response body before it
+            # can be parsed to detect anti-bot methods. This is
+            # necessary because HttpErrorMiddleware is ordered
+            # before HttpCompressionMiddleware per:
+            # 1. https://docs.scrapy.org/en/latest/topics/settings.html#spider-middlewares-base
+            # 2. https://docs.scrapy.org/en/latest/topics/settings.html#downloader-middlewares-base
+            # It is probably not a good idea to reorder these
+            # default middleware orders.
+            httpcm = HttpCompressionMiddleware.from_crawler(spider.crawler)
+            decompressed_response = httpcm.process_response(request, response, spider)
+            html_response = HtmlResponse(decompressed_response.url, body=decompressed_response.body)
+            if html_response.xpath('//meta[@name="description" and @content="px-captcha"]'):
+                self.add_anti_bot_method(AntiBotMethods.HUMAN, spider)
+
+        return response
+
+    @staticmethod
+    def decode_http_header_value(raw_header_value: bytes) -> str:
+        # It's not quite so simple to decode HTTP header values.
```

Comment:
Why not? Have you seen non-unicode values in the headers that you're inspecting?

Reply:
I haven't looked (just relied on clients automatically decoding header values) but imagine UTF-8 is almost universal, and other encodings are esoteric enough to ignore. I had hoped there was some Scrapy method that handles decoding correctly, but haven't found it yet.